docs/Grammar.md

```diff
@@ -155,3 +158,25 @@ untrusted_suffixes = "transition_using_threads" | "propagate_errno"
 
 
 ```
+
+### C-style directives
+
+```c
+
+ifdef_statement:
+    "#ifdef" string statement* [#else statement*] "#endif"
+
+ifndef_statement:
+    "#ifndef" string statement* [#else statement*] "#endif"
+
+statement =
+    include_statement
+    | import_statement
+    | from_import_statement
+    | enum_declaration
+    | struct_declaration
+    | union_declaration
+    | trusted_function
+    | untrusted_function
```

Comment:
This does not capture the following scenario from the PR description:
```c
trusted {
  [region 2: ecall]
  };
```


I think we need to update the grammar to include the directive ..more like

```c
preprocessor_directive :
   "#" "ifdef" identifier
| "#" "ifndef" identifier
| "#" "else"
| "#" "endif"


statement =
   include_statement
    | import_statement
    | from_import_statement
    | enum_declaration
    | struct_declaration
    | union_declaration
    | trusted_function
    | untrusted_function
    | preprocessor_directive

...
trusted_section = "trusted" '{' (trusted_function|preprocessor_directive)* '}' ';'
...
untrusted_section =
    "untrusted" '{' (untrusted_function|preprocessor_directive)* '}'  ';'

untrusted_function =
    function_declaration [unttrusted_suffixes] ';'

...

```

Reply:
I also want to specify the syntax/structure of how "ifdef" should be used.  So I'm thinking something like:

```
include_section: include_statment* [ifdef_include_statement*]
include_statement:  "include" string
ifdef_include_statement: "#" "ifdef" include_statement* ["#" "else" include_statement*] "#" "endif"
```

But this would end up creating ifdef_xxx_statement for other types of statement (import, enum, struct, ... etc). I couldn't find references of parameterized BNF-syntax..

Base on those, I'm a bit lean to having the syntax of preprocessor directive as a sub-section. Since the `trusted_function` and `untrusted_function` are defined as part of the `statement`, it should capture the use of `ifdef`/`ifndef` inside the `trusted section` and `untrusted section`.

Reply:
My concern is that it is not clear how the preprocessor section relates to the rest of the edl language. 
If you were to read the Grammar.md and then look at Parser.cpp, there is no direct mapping.
It is not clear for example, how a `statement` is parsed or how a `statement` relates to `trusted_section` and `untrusted _section`


The C preprocessor grammar for example avoids the notion of a statement:
https://docs.microsoft.com/en-us/cpp/preprocessor/grammar-summary-c-cpp?view=vs-2019

Reply:
Just updated the doc. I update the grammar to make the doc align with the parser.cpp implementation. I also keep the subsection just for referencing the grammar of preprocessor directives themselves.